File: cpp/2593.find-score-of-an-array-after-marking-all-elements.cpp
#include <ios>
#include <iostream>
#include <queue>
#include <unordered_set>
#include <utility>
#include <vector>

using std::vector, std::unordered_set, std::priority_queue, std::pair;

const static int ZERO = []() {
  std::ios_base::sync_with_stdio(false);
  std::cin.tie(nullptr);
  return 0;
}();

class Solution {
public:
  long long findScore(vector<int> &nums) {
    priority_queue<pair<int, int>, vector<pair<int, int>>, std::greater<>> pq;
    for (int i = 0; i < nums.size(); i++) {
      pq.push({nums[i], i});
    }

    unordered_set<int> marked;
    long long score = 0;

    while (!pq.empty()) {
      auto [val, idx] = pq.top();
      pq.pop();

      if (marked.count(idx)) {
        continue;
      }

      score += val;
      marked.insert(idx);
      marked.insert(idx - 1);
      marked.insert(idx + 1);
    }

    return score;
  }
};

File: python/2593.find-score-of-an-array-after-marking-all-elements.py
import heapq


class Solution:
    def findScore(self, nums: list[int]) -> int:
        score = 0
        marked = set()
        hq = [(num, i) for i, num in enumerate(nums)]
        heapq.heapify(hq)

        while hq:
            top, index = heapq.heappop(hq)

            if index in marked:
                continue

            score += top
            marked.add(index)
            marked.add(index + 1)
            marked.add(index - 1)

        return score

